File: kmp_client/iosApp/CommonDI/DobbyConfigsRepositoryImpl.swift
import app
import Foundation

public class DobbyConfigsRepositoryImpl: DobbyConfigsRepository {
    static let shared = DobbyConfigsRepositoryImpl()

    private var userDefaults: UserDefaults = UserDefaults(suiteName: appGroupIdentifier) ?? UserDefaults.standard

    private let cloakConfigKey = "cloakConfigKey"
    private let isCloakEnabledKey = "isCloakEnabledKey"
    private let cloakLocalPortKey = "cloakLocalPortKey"
    private let methodPasswordOutlineKey = "MethodPasswordOutlineKey"
    private let serverPortOutlineKey = "ServerPortOutlineKey"
    private let isOutlineEnabledKey = "isOutlineEnabledKey"
    private let connectionURLKey = "connectionURLKey"
    private let connectionConfigKey = "connectionConfigKey"
    private let prefixOutlineKey = "PrefixOutlineKey"
    private let tcpPathOutlineKey = "TcpPathOutlineKey"
    private let isWebsocketEnabledKey = "isWebsocketEnabledKey"
    private let udpPathOutlineKey = "UdpPathOutlineKey"
    private let isUserInitStopKey = "isUserInitStopKey"

    public func getConnectionURL() -> String {
        return userDefaults.string(forKey: connectionURLKey) ?? ""
    }

    public func setConnectionURL(connectionURL: String) {
        userDefaults.set(connectionURL, forKey: connectionURLKey)
        sync()
    }

    public func getConnectionConfig() -> String {
        return userDefaults.string(forKey: connectionConfigKey) ?? ""
    }

    public func setConnectionConfig(connectionConfig: String) {
        userDefaults.set(connectionConfig, forKey: connectionConfigKey)
        sync()
    }

    public func getCloakConfig() -> String {
        return userDefaults.string(forKey: cloakConfigKey) ?? ""
    }

    public func setCloakConfig(newConfig: String) {
        userDefaults.set(newConfig, forKey: cloakConfigKey)
        sync()
    }

    public func getIsCloakEnabled() -> Bool {
        return userDefaults.bool(forKey: isCloakEnabledKey)
    }

    public func setIsCloakEnabled(isCloakEnabled: Bool) {
        userDefaults.set(isCloakEnabled, forKey: isCloakEnabledKey)
        sync()
    }

    public func getCloakLocalPort() -> Int32 {
        let portValue = userDefaults.object(forKey: cloakLocalPortKey) as? Int ?? 1984
        return Int32(portValue)
    }

    public func setCloakLocalPort(port: Int32) {
        userDefaults.set(Int(port), forKey: cloakLocalPortKey)
        sync()
    }

    public func getServerPortOutline() -> String {
        return userDefaults.string(forKey: serverPortOutlineKey) ?? ""
    }

    public func setServerPortOutline(newConfig: String) {
        userDefaults.set(newConfig, forKey: serverPortOutlineKey)
        sync()
    }

    public func getMethodPasswordOutline() -> String {
        return userDefaults.string(forKey: methodPasswordOutlineKey) ?? ""
    }

    public func setMethodPasswordOutline(newConfig: String) {
        userDefaults.set(newConfig, forKey: methodPasswordOutlineKey)
        sync()
    }

    public func getIsOutlineEnabled() -> Bool {
        return userDefaults.bool(forKey: isOutlineEnabledKey)
    }

    public func setIsOutlineEnabled(isOutlineEnabled: Bool) {
        userDefaults.set(isOutlineEnabled, forKey: isOutlineEnabledKey)
        sync()
    }

    public func getPrefixOutline() -> String {
        return userDefaults.string(forKey: prefixOutlineKey) ?? ""
    }

    public func setPrefixOutline(prefix: String) {
        userDefaults.set(prefix, forKey: prefixOutlineKey)
        sync()
    }

    public func getTcpPathOutline() -> String {
        return userDefaults.string(forKey: tcpPathOutlineKey) ?? ""
    }

    public func setTcpPathOutline(tcpPath: String) {
        userDefaults.set(tcpPath, forKey: tcpPathOutlineKey)
        sync()
    }

    public func getIsWebsocketEnabled() -> Bool {
        return userDefaults.bool(forKey: isWebsocketEnabledKey)
    }

    public func setIsWebsocketEnabled(enabled: Bool) {
        userDefaults.set(enabled, forKey: isWebsocketEnabledKey)
        sync()
    }

    public func getUdpPathOutline() -> String {
        return userDefaults.string(forKey: udpPathOutlineKey) ?? ""
    }

    public func setUdpPathOutline(udpPath: String) {
        userDefaults.set(udpPath, forKey: udpPathOutlineKey)
        sync()
    }

    public func getAwgConfig() -> String {
        return ""
    }

    public func getIsAmneziaWGEnabled() -> Bool {
        return false
    }

    public func getVpnInterface() -> VpnInterface {
        return VpnInterface.cloakOutline
    }

    public func setAwgConfig(newConfig: String) {}

    public func setIsAmneziaWGEnabled(isAmneziaWGEnabled: Bool) {}

    public func setVpnInterface(vpnInterface: VpnInterface) {}

    public func couldStart() -> Bool {
        return true
    }

    public func getIsUserInitStop() -> Bool {
        return userDefaults.bool(forKey: isUserInitStopKey)
    }

    public func setIsUserInitStop(isUserInitStop: Bool) {
        userDefaults.set(isUserInitStop, forKey: isUserInitStopKey)
        sync()
    }

    public func sync() {
        userDefaults.synchronize()
    }
}

File: kmp_client/app/src/androidMain/kotlin/com/dobby/domain/DobbyConfigsRepositoryImpl.kt
package com.dobby.domain

import android.content.SharedPreferences
import com.dobby.feature.main.domain.DobbyConfigsRepository
import com.dobby.feature.main.domain.VpnInterface
import android.util.Log.i as AndroidLog
import com.dobby.outline.OutlineGo

internal class DobbyConfigsRepositoryImpl(
    private val prefs: SharedPreferences
) : DobbyConfigsRepository {
    override fun getVpnInterface(): VpnInterface {
        val prefsResult = prefs.getString("vpnInterface", VpnInterface.DEFAULT_VALUE.toString())
            ?: VpnInterface.DEFAULT_VALUE.toString()
        AndroidLog("DOBBY_TAG", "getVpnInterface: $prefsResult")

        return VpnInterface.valueOf(prefsResult)
    }

    override fun setVpnInterface(vpnInterface: VpnInterface) {
        prefs.edit().putString("vpnInterface", vpnInterface.toString()).apply().also {
            AndroidLog("DOBBY_TAG", "setVpnInterface: $vpnInterface")
        }
    }

    override fun getConnectionURL(): String {
        return (prefs.getString("сonnectionURL", "") ?: "").also {
            AndroidLog("DOBBY_TAG", "getConnectionURL, url = ${it}")
        }
    }

    override fun setConnectionURL(connectionURL: String) {
        prefs.edit().putString("сonnectionURL", connectionURL).apply().also {
            AndroidLog("DOBBY_TAG", "setConnectionURL, url = ${connectionURL}")
        }
    }

    override fun getCloakConfig(): String {
        return (prefs.getString("cloakConfig", "") ?: "").also {
            AndroidLog("DOBBY_TAG", "getCloakConfig, size = ${it.length}")
        }
    }

    override fun setCloakConfig(newConfig: String) {
        prefs.edit().putString("cloakConfig", newConfig).apply().also {
            AndroidLog("DOBBY_TAG", "setCloakConfig, size = ${newConfig.length}")
        }
    }

    override fun getConnectionConfig(): String {
        return (prefs.getString("сonnectionConfig", "") ?: "").also {
            AndroidLog("DOBBY_TAG", "getConnectionConfig, config = ${it}")
        }
    }

    override fun setConnectionConfig(connectionConfig: String) {
        prefs.edit().putString("сonnectionConfig", connectionConfig).apply().also {
            AndroidLog("DOBBY_TAG", "setConnectionConfig, config = ${connectionConfig}")
        }
    }

    override fun getIsCloakEnabled(): Boolean {
        return prefs.getBoolean("isCloakEnabled", false).also {
            AndroidLog("DOBBY_TAG", "getIsCloakEnabled: $it")
        }
    }

    override fun setIsCloakEnabled(isCloakEnabled: Boolean) {
        prefs.edit().putBoolean("isCloakEnabled", isCloakEnabled).apply().also {
            AndroidLog("DOBBY_TAG", "setIsCloakEnabled: $isCloakEnabled")
        }
    }

    override fun getCloakLocalPort(): Int {
        return prefs.getInt("cloakLocalPort", 1984).also {
            AndroidLog("DOBBY_TAG", "getCloakLocalPort: $it")
        }
    }

    override fun setCloakLocalPort(port: Int) {
        prefs.edit().putInt("cloakLocalPort", port).apply().also {
            AndroidLog("DOBBY_TAG", "setCloakLocalPort: $port")
        }
    }

    override fun setServerPortOutline(newConfig: String) {
        prefs.edit().putString("ServerPortOutlineKey", newConfig).apply().also {
            AndroidLog("DOBBY_TAG", "setServerPortOutline, size = ${newConfig.length}")
        }
    }

    override fun setMethodPasswordOutline(newConfig: String) {
        prefs.edit().putString("MethodPasswordOutlineKey", newConfig).apply().also {
            AndroidLog("DOBBY_TAG", "setMethodPasswordOutline, size = ${newConfig.length}")
        }
    }

    override fun getServerPortOutline(): String {
        return (prefs.getString("ServerPortOutlineKey", "") ?: "").also {
            AndroidLog("DOBBY_TAG", "getServerPortOutline, size = ${it.length}")
        }
    }

    override fun getMethodPasswordOutline(): String {
        return (prefs.getString("MethodPasswordOutlineKey", "") ?: "").also {
            AndroidLog("DOBBY_TAG", "getMethodPasswordOutline, size = ${it.length}")
        }
    }

    override fun getIsOutlineEnabled(): Boolean {
        return prefs.getBoolean("isOutlineEnabled", false).also {
            AndroidLog("DOBBY_TAG", "getIsOutlineEnabled = $it")
        }
    }

    override fun setIsOutlineEnabled(isOutlineEnabled: Boolean) {
        prefs.edit().putBoolean("isOutlineEnabled", isOutlineEnabled).apply().also {
            AndroidLog("DOBBY_TAG", "setIsOutlineEnabled = $isOutlineEnabled")
        }
    }

    override fun getPrefixOutline(): String {
        return (prefs.getString("PrefixOutlineKey", "") ?: "").also {
            AndroidLog("DOBBY_TAG", "getPrefixOutline, value = $it")
        }
    }

    override fun setPrefixOutline(prefix: String) {
        prefs.edit().putString("PrefixOutlineKey", prefix).apply().also {
            AndroidLog("DOBBY_TAG", "setPrefixOutline, value = $prefix")
        }
    }

    override fun getTcpPathOutline(): String {
        return (prefs.getString("TcpPathOutlineKey", "") ?: "").also {
            AndroidLog("DOBBY_TAG", "getTcpPathOutline, value = $it")
        }
    }

    override fun setTcpPathOutline(tcpPath: String) {
        prefs.edit().putString("TcpPathOutlineKey", tcpPath).apply().also {
            AndroidLog("DOBBY_TAG", "setTcpPathOutline, value = $tcpPath")
        }
    }

    override fun getIsWebsocketEnabled(): Boolean {
        return prefs.getBoolean("isWebsocketEnabled", false).also {
            AndroidLog("DOBBY_TAG", "getIsWebsocketEnabled = $it")
        }
    }

    override fun setIsWebsocketEnabled(enabled: Boolean) {
        prefs.edit().putBoolean("isWebsocketEnabled", enabled).apply().also {
            AndroidLog("DOBBY_TAG", "setIsWebsocketEnabled = $enabled")
        }
    }

    override fun getUdpPathOutline(): String {
        return (prefs.getString("UdpPathOutlineKey", "") ?: "").also {
            AndroidLog("DOBBY_TAG", "getUdpPathOutline, value = $it")
        }
    }

    override fun setUdpPathOutline(udpPath: String) {
        prefs.edit().putString("UdpPathOutlineKey", udpPath).apply().also {
            AndroidLog("DOBBY_TAG", "setUdpPathOutline, value = $udpPath")
        }
    }

    override fun getAwgConfig(): String {
        return (prefs.getString("awgConfig", DEFAULT_AWG_CONFIG) ?: "").also {
            AndroidLog("DOBBY_TAG", "getAwgConfig, size = ${it.length}")
        }
    }

    override fun setAwgConfig(newConfig: String) {
        prefs.edit().putString("awgConfig", newConfig).apply().also {
            AndroidLog("DOBBY_TAG", "setAwgConfig, size = ${newConfig.length}")
        }
    }

    override fun getIsAmneziaWGEnabled(): Boolean {
        return prefs.getBoolean("isAmneziaWGEnabled", false).also {
            AndroidLog("DOBBY_TAG", "getIsAmneziaWGEnabled = $it")
        }
    }

    override fun setIsAmneziaWGEnabled(isAmneziaWGEnabled: Boolean) {
        prefs.edit().putBoolean("isAmneziaWGEnabled", isAmneziaWGEnabled).apply().also {
            AndroidLog("DOBBY_TAG", "setIsAmneziaWGEnabled = $isAmneziaWGEnabled")
        }
    }

    override fun couldStart(): Boolean {
        return true
    }

    override fun getIsUserInitStop(): Boolean {
        return prefs.getBoolean("isUserInitStop", true).also {
            AndroidLog("DOBBY_TAG", "getIsUserInitStop = $it")
        }
    }

    override fun setIsUserInitStop(isUserInitStop: Boolean) {
        prefs.edit().putBoolean("isUserInitStop", isUserInitStop).apply().also {
            AndroidLog("DOBBY_TAG", "setIsUserInitStop = $isUserInitStop")
        }
    }

    companion object {
        const val DEFAULT_AWG_CONFIG = """[Interface]
PrivateKey = <...>
Address = <...>
DNS = 8.8.8.8
Jc = 0
Jmin = 0
Jmax = 0
S1 = 0
S2 = 0
H1 = 1
H2 = 2
H3 = 3
H4 = 4

[Peer]
PublicKey = <...>
Endpoint = <...>
AllowedIPs = 0.0.0.0/0
PersistentKeepalive = 60
"""
    }
}
